objectnode/api_handler_object.go

```diff
@@ -357,18 +389,49 @@ func (o *ObjectNode) headObjectHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// set response header
-	if len(fileInfo.ETag) > 0 {
-		w.Header().Set(HeaderNameETag, wrapUnescapedQuot(fileInfo.ETag))
-	}
 	w.Header().Set(HeaderNameAcceptRange, HeaderValueAcceptRange)
+	w.Header().Set(HeaderNameLastModified, formatTimeRFC1123(fileInfo.ModifyTime))
+	w.Header().Set(HeaderNameContentMD5, EmptyContentMD5String)
 	if len(fileInfo.MIMEType) > 0 {
 		w.Header().Set(HeaderNameContentType, fileInfo.MIMEType)
 	} else {
 		w.Header().Set(HeaderNameContentType, HeaderValueTypeStream)
 	}
-	w.Header().Set(HeaderNameLastModified, formatTimeRFC1123(fileInfo.ModifyTime))
-	w.Header().Set(HeaderNameContentLength, strconv.Itoa(int(fileInfo.Size)))
-	w.Header().Set(HeaderNameContentMD5, EmptyContentMD5String)
+
+	// check request is whether contain param : partNumber
+	partNumber := r.URL.Query().Get(ParamPartNumber)
+	if len(partNumber) > 0 && fileInfo.Size >= MinParallelDownloadFileSize {
+		partNumberInt, err := strconv.ParseUint(partNumber, 10, 64)
+		if err != nil {
+			log.LogErrorf("getObjectHandler: parse param partNumber(%s) fail: requestID(%v) err(%v)", partNumber, GetRequestID(r), err)
+			errorCode = InvalidArgument
+			return
+		}
+		partSize, partCount, rangeLower, rangeUpper, err := parsePartInfo(partNumberInt, uint64(fileInfo.Size))
+		log.LogDebugf("headObjectHandler: parsed partSize(%d), partCount(%d), rangeLower(%d), rangeUpper(%d)", partSize, partCount, rangeLower, rangeUpper)
+		if err != nil {
+			errorCode = InternalErrorCode(err)
+			return
+		}
+		if partNumberInt > partCount{
+			log.LogErrorf("getObjectHandler: param partNumber(%d) is more then partCount: requestID(%v)", partNumberInt, partCount, GetRequestID(r))
+			errorCode = NoSuchKey
+			return
+		}
+		w.Header().Set(HeaderNameContentLength, strconv.Itoa(int(partSize)))
+		w.Header().Set(HeaderNameContentRange, fmt.Sprintf("bytes %d-%d/%d", rangeLower, rangeUpper, fileInfo.Size))
+		w.Header().Set(HeaderNameDownloadPartCount, strconv.Itoa(int(partCount)))
+		if len(fileInfo.ETag) > 0 && !strings.Contains(fileInfo.ETag, "-") {
+			w.Header().Set(HeaderNameETag, fmt.Sprintf("%s-%d", fileInfo.ETag, partCount))
+		}
+	} else {
+		w.Header().Set(HeaderNameContentLength, strconv.Itoa(int(fileInfo.Size)))
+		if len(fileInfo.ETag) > 0 {
+			log.LogDebugf("ETag : %s", wrapUnescapedQuot(fileInfo.ETag))
```

Comment:
Pay attention to whether there are unnecessary logs, and remove them if they exist.